### PR TITLE
[stable/gocd] Add support for SSH keys on server and agent

### PR DESCRIPTION
**Pull request by or1can**
This PR adds support for setting SSH keys on the server and agent. This allows, for example, checking out a git repository via SSH rather than HTTPS.

### CODE REVIEW

**Comment by varshavaradarajan** [comment]
@or1can - thank you for this PR! I think that its a great improvement on the GoCD Helm chart. Do you mind rebasing and signing the CNCF CLA so that we can review and get this merged?

**Comment by cfontes** [comment]
Hello there, this is super useful, can I help to get this merged in anyway?

**Comment by or1can** [comment]
That took me way longer than planned to get the CNCF CLA signed. I've just rebased, so hopefully it's good to go now.

**Comment by varshavaradarajan** [comment]
@or1can - still shows that the CLA hasn't been signed.

**Comment by varshavaradarajan** [comment]
And please check circle ci console and fix the tests.

**Comment by cfontes** [comment]
I am testing this locally and it looks like the files are created under root ownership, because of that my GoCD server is not able to use the credentials in Github.

Does it work for you guys?

**Comment by or1can** [comment]
yeah, I have a simple pipeline that polls a Github repo and builds a job, which seems to work well.

The only issue I've seen so far is with the elastic agent, which doesn't include the ssh secret by default. To replicate the elastic agent demo-app profile, you need to add a new elastic agent profile with the following Pod Yaml to mount the ssh secret in the correct location and then configure the job to use this new profile:

```
apiVersion: v1
kind: Pod
metadata:
  name: gocd-ea-{{ POD_POSTFIX }}
  labels:
    app: web
spec:
  volumes:
    - name: ssh-secrets
      secret:
        secretName: gocd-ssh
  containers:
    - name: gocd-agent-docker-dind-{{ CONTAINER_POSTFIX }}
      image: gocd/gocd-agent-docker-dind:v18.2.0
      securityContext:
        privileged: true
      volumeMounts:
        - name: ssh-secrets
          readOnly: true
          mountPath: /home/go/.ssh
```

I'll update the PR to include this in the README.

**Comment by cfontes** [comment]
Is your container running as root?

**Comment by or1can** [comment]
Also, for clarity, we're using the good yaml config plugin to define pipelines.

A proof of concept configuration which works for us is as follows (change the git and elastic_profile_id values as appropriate):

```
format_version: 2
pipelines:
  mypipe1:
    group: mygroup1
    materials:
      mygit:
        git: git@github.com....
    stages:
      - build:
          jobs:
            build:
              elastic_profile_id: "some-elasticagent-profile-id"
              tasks:
                - exec:
                    command: echo
                    arguments:
                      - "Fake test run"
```

File: stable/gocd/README.md [hunk 1]
```diff
@@ -214,6 +214,33 @@ $ helm install --name gocd-app --set server.persistence.existingClaim=PVC_NAME s
...
+structured to hold the entire contents of the .ssh folder on the GoCD server and agent.
+
+```bash
+ssh-keygen -t rsa -b 4096 -C "user@example.com" -f gocd-ssh -P ''
+
+ssh-keyscan [host] > gocd_known_hosts
+
+kubectl create secret generic gocd-ssh \
+    --from-file=id-rsa=gocd-ssh \
+    --from-file=id-rsa.pub=gocd-ssh.pub \
```
Comment (akshaydewan) [inline stable/gocd/README.md:234]: @or1can  Could you change `id-rsa` to `id_rsa` (with an underscore) for the private and public keys? The ssh client picks up names with underscores by default, and the hyphen might confuse users.

**Comment by varshavaradarajan** [comment]
@or1can - please update the PR when you can.

**Comment by varshavaradarajan** [comment]
@or1can - did you check that the known-hosts file also must be provided?

**Comment by varshavaradarajan** [comment]
@or1can - something went wrong with the rebase? Please update the PR with only your changes to gocd. Thanks! :)

**Comment by cfontes** [comment]
Hello guys... any updates on this? 
It would be super helpful to have it merged. I am maintaining my own copy of it locally because right now but it is kind of a pain and makes this helm kind of hard to use not having it.

Can I do something to help? Open another PR or something?

**Comment by GaneshSPatil** [comment]
Hello @or1can -- any updates on the PR?

**Comment by lb-saildrone** [comment]
Hi all, I ran into the same issue here but was able to get around it by using the kubernetes api's existing features to inject my ssh key into the right volumes before gocd launches.  these patches can be applied on top of an existing helm deployment and have the following prequisites:
1) I used the following versions, I believe most of this should work on 1.6 and up but i'm not 100% sure:
```
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.3", GitCommit:"2bba0127d85d5a46ab4b778548be28623b32d0b0", GitTreeState:"clean", BuildDate:"2018-05-28T20:16:17Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.3", GitCommit:"2bba0127d85d5a46ab4b778548be28623b32d0b0", GitTreeState:"clean", BuildDate:"2018-05-28T20:13:43Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
```
2) a helm deployment of gocd running in the 'gocd' namespace (you can use whatever namespace you like, just be sure to change it).
3) a config map in the same namespace called gocd-ssh-config containing the contents of .ssh/config:
```
 # I ran this command to upload the contents of 
 $ kubectl create configmap gocd-ssh-config --from-file=gocd/ssh_config -n gocd

# ssh_config contents
    Host github.com
        StrictHostKeyChecking no
        UserKnownHostsFile /dev/null
        IdentityFile /home/go/.ssh/id_rsa
```
4) a secret in the same namespace called github ssh that contains the public/private key you are using to ssh auth into github. I ran this command to create the secret from existing files:
```
 kubectl create secret generic github-ssh --from-file=./secrets/id_rsa --from-file=./secrets/id_rsa.pub --namespace gocd
```
Now you can apply the following patch using `kubectl patch deployment gocd-server --patch "$(cat gocd/patch-ssh-secret-server.yaml)" -n gocd`
```
# gocd/patch-ssh-secret-server.yaml
spec:
  template:
    spec:
      containers:
      - name: gocd-server
        lifecycle:
          postStart:
            exec:
              command:
              - /bin/bash
              - /preconfigure_server.sh
      initContainers:
      - args:
        - |-
          while [ ! -e /home/go/ssh_config -o ! -e /home/go/ssh_key ]; do sleep 1; done &&
          mkdir -p /home/go/.ssh &&
          cp /home/go/ssh_config/ssh_config /home/go/.ssh/config &&
          cp /home/go/ssh_key/* /home/go/.ssh/ &&
          chmod 400 /home/go/.ssh/id_rsa* &&
          chown -R go /home/go/.ssh/
        command:
        - /bin/bash
        - -c
        image: gocd/gocd-server:v18.6.0
        imagePullPolicy: IfNotPresent
        name: ssh-config
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /home/go
          name: goserver-vol
          subPath: homego
        - mountPath: /home/go/ssh_config
          name: ssh-config
        - mountPath: /home/go/ssh_key
          name: ssh-key
      volumes:
      - configMap:
          defaultMode: 420
          name: gocd-ssh-config
        name: ssh-config
      - name: ssh-key
        secret:
          defaultMode: 420
          secretName: github-ssh
```
YMMV but this got the ssh key working on the server. I used a similar technique to configure the kubernetes elastic agent from within the ui I went to admin->elastic agent profiles created my profile and added the following sections to the spec:
```
  initContainers:
  - name: ssh-config
    image: gocd/gocd-agent-docker-dind:v18.6.0
    imagePullPolicy: IfNotPresent
    command:
    - /bin/bash
    - -c
    args:
    - |-
      while [ ! -e /home/go/ssh_config -o ! -e /home/go/ssh_key ]; do sleep 1; done &&
      mkdir -p /home/go/.ssh &&
      cp /home/go/ssh_config/ssh_config /home/go/.ssh/config &&
      cp /home/go/ssh_key/* /home/go/.ssh/ &&
      chmod 400 /home/go/.ssh/id_rsa* &&
      chown -R go /home/go/.ssh/
    volumeMounts:
    - mountPath: /home/go
      name: goagent-vol
      subPath: homego
    - mountPath: /home/go/ssh_config
      name: ssh-config
    - mountPath: /home/go/ssh_key
      name: ssh-key
 # containers config, etc.
 # ....
  volumes:
  - configMap:
      defaultMode: 420
      name: gocd-ssh-config
    name: ssh-config
  - name: ssh-key
    secret:
      defaultMode: 420
      secretName: github-ssh
  - name: goagent-vol
    emptyDir: {}kk
```

**Comment by MPV** [comment]
Sweet @lb-saildrone, might give it a try. Though it would be even smoother (especially for Elastic Agents) to get the originally suggested change incorporated.

@or1can, is this the only commit needed? ca41a5b885642d0f560016b437ca81817d440658

Maybe we could create a new pull request with just this commit to start of from again?
Seems like you might've accidentally updated your own `master` branch (that this PR was based on) with newer commits from upstream.

**Comment by lb-saildrone** [comment]
@MPV If I am groking this correctly, you are still going to have to configure the elastic agent from within gocd once it is launched (or pumping in the right XML before launch).  This helm chart only configures and installs the static agent but not the elastic ones.  I'm currently working on a new PR to incorporate these additions (and possibly solve the EA problem) and will submit it against incubator in the coming week.

**Comment by varshavaradarajan** [comment]
@lb-saildrone - couple of things

The GoCD helm chart is in stable. The incubator chart is deprecated. I'd suggest submitting PRs for stable itself.

The GoCD helm chart comes with the Kubernetes elastic agent installed and preconfigured to use the same k8s cluster. Check out the configmap yaml in the templates directory. If you think that the elastic agent profile values could be improved in a generic sense, feel free to create a PR. 

Thanks!

**Comment by lb-saildrone** [comment]
@varshavaradarajan sure, my only point is that this PR (as far as I can tell) does not do anything to add the ability to configure ssh into the EA profile. As such that needs to either be added to the preconfigure_server.sh script or it needs to be added manually after the fact. If the EA does not have access to the ssh key, it will never be able to ssh-auth into github (i.e. add ssh support)

**Comment by MPV** [comment]
@lb-saildrone you're right. I ended up going with the elastic agent profile method for configuring ssh & git instead.

**Comment by MPV** [comment]
Also, this PR seems to now be superseded by https://github.com/helm/charts/pull/7084

**Comment by Dilip-Nandakumar** [comment]
@MPV i created a new PR so we can have the changes incorporated sooner as the this PR has run into some issues and there were no updates for quite some time.

Like @MPV and @lb-saildrone mentioned this only works for static agents. For elastic gocd agents to work i had to create a new elastic agent profile in GoCD. 

I have mentioned the same in the new PR 

https://github.com/helm/charts/pull/7084#issuecomment-412018277

**Comment by GaneshSPatil** [comment]
Closing this PR in favour of https://github.com/helm/charts/pull/7084
/close

**Comment by MPV** [comment]
I think this can be closed now @or1can since #7084 is merged instead. 😊

**Pull request closed without merging**